File: src/main/java/org/telran/lecture_12_trees/practice/BST.java
package org.telran.lecture_12_trees.practice;

class Node {
    int key;
    String value;
    Node left;
    Node right;

    /**
     * Конструктор для создания нового узла.
     *
     * @param key   Ключ узла.
     * @param value Значение узла.
     */
    public Node(int key, String value) {
        this.key = key;
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class BST {
    Node root;

    /**
     * Конструктор для создания нового бинарного дерева поиска.
     */
    public BST() {
        this.root = null;
    }

    /**
     * Вставляет новый узел в дерево.
     *
     * @param key   Ключ нового узла.
     * @param value Значение нового узла.
     */
    public void insert(int key, String value) {
        Node newNode = new Node(key, value);
        if (root == null) {
            root = newNode;
        } else {
            insertNode(root, newNode);
        }
    }

    /**
     * Рекурсивно вставляет новый узел в поддерево.
     *
     * @param node    Текущий узел.
     * @param newNode Новый узел.
     */
    private void insertNode(Node node, Node newNode) {
        // TODO-1: Доработайте код, чтобы все ключи были уникальные.
        if (newNode.key == node.key) {
            node.value = newNode.value;
            return;
        }
        if (newNode.key < node.key) {
            if (node.left == null) {
                node.left = newNode;
            } else {
                insertNode(node.left, newNode);
            }
        } else {
            if (node.right == null) {
                node.right = newNode;
            } else {
                insertNode(node.right, newNode);
            }
        }
    }

    /**
     * Ищет значение узла по заданному ключу.
     *
     * @param key Ключ искомого узла.
     * @return Значение найденного узла или null, если узел не найден.
     */
    public String search(int key) {
        Node node = searchNode(root, key);
        return node != null ? node.value : null;
    }

    /**
     * Рекурсивно ищет узел с заданным ключом в поддереве.
     *
     * @param node Текущий узел.
     * @param key  Ключ искомого узла.
     * @return Найденный узел или null, если узел не найден.
     */
    private Node searchNode(Node node, int key) {
        // TODO-3: напишите реализацию метода
        if (node == null) {
            return null;
        }
        if (node.key == key) {
            return node;
        } else if (key < node.key) {
            return searchNode(node.left, key);
        } else {
            return searchNode(node.right, key);
        }

        // Если узел найден, возвращаем его значение(value)
        // Если не найден, продолжаем поиск в левой или правой части

    }

    /**
     * Находит минимальный узел в дереве.
     *
     * @return Минимальный узел или null, если дерево пустое.
     */
    public Node min() {
        // TODO-4: напишите реализацию метода
        if (root == null) {
            return null;
        }
        Node current = root;
        while (current.left != null) {
            current = current.left;
        }
        return current;
    }

    /**
     * Находит максимальный узел в дереве.
     *
     * @return Максимальный узел или null, если дерево пустое.
     */
    public Node max() {
        // TODO-5: напишите реализацию метода
        if (root == null) {
            return null;
        }
        Node current = root;
        while (current.right != null) {
            current = current.right;
        }
        return current; // Заглушка
    }

    public int length() {
        // TODO-2: реализуйте метод, возвращающий количество узлов в дереве
        return length(root);
    }

    private int length(Node node) {
        if (node == null) {
            return 0;
        }
        return 1 + length(node.left) + length(node.right);
    }

    public int height() {
        return height(root);
    }

    private int height(Node node) {
        if (node == null) {
            return 0;
        }
        return 1 + Math.max(height(node.left), height(node.right)); // Берём макс. глубину левого и правого поддерева
    }

    /**
     * Отображает дерево в виде древовидной структуры в терминале.
     */
    public void displayTree() {
        if (root == null) {
            System.out.println("Дерево пустое.");
            return;
        }
        displayNode(root, "", true);
    }

    /**
     * Рекурсивно отображает узел и его поддеревья.
     *
     * @param node   Текущий узел.
     * @param prefix Префикс для отображения уровня дерева.
     * @param isLeft Флаг, указывающий, является ли узел левым дочерним.
     */
    private void displayNode(Node node, String prefix, boolean isLeft) {
        if (node != null) {
            String displayPrefix = prefix + (isLeft ? "├── " : "└── ");
            String directionIndicator = isLeft ? "L: " : "R: ";

            System.out.println(displayPrefix + directionIndicator + node.key + ":" + node.value);

            String childPrefix = prefix + (isLeft ? "│   " : "    ");
            if (node.left != null || node.right != null) {
                displayNode(node.left, childPrefix, true);
                displayNode(node.right, childPrefix, false);
            }
        }
    }

    public static void main(String[] args) {

        BST bst = new BST();

        bst.insert(6, "раз");
        bst.insert(3, "два");
        bst.insert(8, "три");
        bst.insert(2, "four");
        bst.insert(9, "пять");
        bst.insert(4, "six");
        bst.insert(45, "seven");
        bst.insert(63, "eight");
        bst.insert(6, "nine");
        bst.insert(-10, "ten");
        bst.insert(-4, "eleven");
        bst.insert(23, "twelve");
        bst.insert(19, "thirteen");

        bst.displayTree();

        System.out.println("Search of 7: " + (bst.search(7) != null ? bst.search(7) : "Not Found"));
        System.out.println("Search of 100: " + (bst.search(100) != null ? bst.search(100) : "Not Found"));
        System.out.println("Search of 45: " + (bst.search(45) != null ? bst.search(45) : "Not Found"));

        System.out.println("---------------");

        System.out.println("Минимальный узел: " + bst.min().key + " (" + bst.min().value + ")");
        System.out.println("---------------");
        System.out.println("Максимальный узел: " + bst.max().key + " (" + bst.max().value + ")");
        System.out.println("---------------");
        System.out.println("Количество узлов в дереве составляет: " + bst.length());
        System.out.println("---------------");
        System.out.println("Глубина или высота дерева: " + bst.height());
    }
}

File: src/main/java/org/telran/lecture_12_trees/practice/BST.js
/**
 * Класс, представляющий узел бинарного дерева поиска.
 */
class Node {
    /**
     * Создает новый узел.
     * @param key Ключ узла.
     * @param value Значение узла.
     */
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

/**
 * Класс, представляющий бинарное дерево поиска.
 */
class BinarySearchTree {
    /**
     * Создает новое бинарное дерево поиска.
     */
    #length=0
    constructor(root, length) {
        this.root = null;
        this.#length=this.#length
    }

    /**
     * Вставляет новый узел в дерево.
     * @param key Ключ нового узла.
     * @param value Значение нового узла.
     */
    insert(key, value) {
        const newNode = new Node(key, value);

        if (this.root === null) {
            this.root = newNode;
            this.#length++;
        } else {
            this.insertNode(this.root, newNode);
        }
    }

    /**
     * Рекурсивно вставляет новый узел в поддерево.
     * @param node Текущий узел.
     * @param newNode Новый узел.
     */
    insertNode(node, newNode) {
        // TODO-1: Доработайте код, чтобы все ключи были уникальные.
        if (newNode.key < node.key) {
            if (node.left === null) {
                node.left = newNode;
            } else {
                this.insertNode(node.left, newNode);
                this.#length++
            }
        } 
        if (newNode.key > node.key) {
            if (node.left === null) {
                node.left = newNode;
            } else {
                this.insertNode(node.left, newNode);
                this.#length++
            }
        }
        if (newNode.key === node.key){
            console.log(`"Ошибка, дублированный ключ ${newNode.key}"`);
            
        }
    }

    /**
     * Ищет значение узла по заданному ключу.
     * @param key Ключ искомого узла.
     * @returns Значение найденного узла или null, если узел не найден.
     */
    search(key) {
        const node = this.searchNode(this.root, key);
        return node ? node.value : null;
    }

    /**
     * Рекурсивно ищет узел с заданным ключом в поддереве.
     * @param node Текущий узел.
     * @param key Ключ искомого узла.
     * @returns Найденный узел или null, если узел не найден.
     */
    searchNode(node, key) {
        // // TODO-3:  напишите реализацию метода

        // Если узел найден, возвращаем его значение(value)
        // Если не найден, продолжаем поиск в левой или правой части
        if (node === null) {
        return null;
      }
  
        if (node.key === key) {
        return node; 
      }
  
        if (key < node.key) {
        return this.searchNode(node.left, key); 
      }
        if (key > node.key) {
        return this.searchNode(node.right, key);
    }
    }

    /**
     * Находит минимальный узел в дереве.
     * @returns Минимальный узел или null, если дерево пустое.
     * // TODO-4: напишите реализацию метода
     */
    min() {
        if (this.root === null) {
            return null;
        }

        let current = this.root;

        while (current.left !== null) {
            current = current.left;
        }

        return current; 
    }

    /**
     * Находит максимальный узел в дереве.
     * @returns Максимальный узел или null, если дерево пустое.
     *  // TODO-5: напишите реализацию метода
     */
    max() {
        if (this.root === null) {
            return null;
        }

        let current = this.root;

        while (current.right !== null) {
            current = current.right;
        }
        return current; 
    }

    // TODO-2: реализуйте метод, возвращающий количество узлов в дереве
    // length() {
        
    //     const countNodes = (node) => {
    //         if (node === null) {
    //             return 0; 
    //         }
    //         return 1 + countNodes(node.left) + countNodes(node.right);
    //     };        
    //     return countNodes(this.root);        
    // }

    length() {
        return this.#length;
    }

    /**
     * Отображает дерево в виде древовидной структуры в терминале.
     */
    displayTree() {
        if (this.root === null) {
            console.log('Дерево пустое.');
            return;
        }

        this.displayNode(this.root, '', true);
    }

    /**
     * Рекурсивно отображает узел и его поддеревья.
     * @param node Текущий узел.
     * @param prefix Префикс для отображения уровня дерева.
     * @param isLeft Флаг, указывающий, является ли узел левым дочерним.
     */
    displayNode(node, prefix, isLeft) {
        if (node !== null) {
            const displayPrefix = prefix + (isLeft ? '├── L:' : '└── R:');
            // console.log(displayPrefix + node.key + ':' + node.value);
            console.log(displayPrefix + node.key);

            const childPrefix = prefix + (isLeft ? '│   ' : '    ');
            if (node.left || node.right) {
                this.displayNode(node.left, childPrefix, true);
                this.displayNode(node.right, childPrefix, false);
            }
        }
    }
}

// Пример использования
const bts = new BinarySearchTree();
bts.insert(1, "1");
bts.insert(3, "1");
bts.insert(4, "1");
bts.insert(2, "1");
bts.insert(6, "1");
bts.insert(8, "1");
bts.insert(5, "1");

bts.displayTree();

console.log(bts.length());
